Uniform buffers for shader props, allow variable number of channels, update deck.gl to 9.2.9

xinaesthete opened the pull request:
Fixes #687
<!-- For other PRs without open issue -->
#### Background

https://github.com/hms-dbmi/viv/issues/687 as well as elsewhere for discussion of UBO etc.

This is work-in-progress... started out just trying to update to latest luma.gl/deck.gl to deal with some peer-dependency issues downstream. Latest versions remove the deprecated `setUniforms` etc, so this is becoming more pressing.

This is still WIP as of this writing, but I think starting to take shape for a way of supporting N-channels without too much pain for extension authors.

#### Change List
- declaring shader uniforms as UBO
- setting props with `model.shaderInputs` rather than `setUniforms`
- tending to prefer implementing `updateState()` rather than `draw()`, I think this is more in line with how things are supposed to be
- trying to make sure any assumptions about `MAX_CHANNELS = 6` is findable in the code etc...
- shaders have a define for `NUM_CHANNELS` which should be used consistently where relevant
- initial version of a shader code preprocessor wrapped in a new `VivShaderAssembler` class for expanding lines with `<VIV_CHANNEL_INDEX>`...
#### Checklist
 - [x] Update [JSdoc types](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) if there is any API change.
 - [x] Make sure Avivator works as expected with your change.
 - [x] Update documentation for migrating shader extensions to the new form once the design is settled.

xinaesthete committed:
update deck and luma versions, with changes to make sure pnpm i is clean.
some tests are failing with this revision, runtime not tried.

xinaesthete committed:
Revert "update deck and luma versions, with changes to make sure pnpm i is clean."

This reverts commit c51bffee0ab79cd49b108329654b2f1410cfa314.

xinaesthete committed:
refactor additive-colormap-extension to UBO

xinaesthete committed:
change BaseExtension draw to updateState

xinaesthete committed:
rewrite shaders to use NUM_CHANNELS (still very much fixed) Max channels allowed #687

xinaesthete committed:
move NUM_CHANNELS define to default module

xinaesthete committed:
more consistent use of NUM_CHANNELS, some comments

xinaesthete committed:
add VivShaderAssembler, fix volume renderer missing MAX_CHANNELS

xinaesthete committed:
first pass at shader-preprocess

dbmi-svc-checkmarx commented:
![Logo](https://cdn.ast.checkmarx.net/integrations/logo/Checkmarx.png)
**Checkmarx One – Scan Summary & Details** – [765f8662\-47a8\-4732\-aaf2\-57ca5c3616aa](https://us.ast.checkmarx.net/projects/0b20d2a1-41f2-4435-88e0-f87103d01fc9/scans?id=765f8662-47a8-4732-aaf2-57ca5c3616aa&branch=xinaesthete%2Fdeck_9.2)
<!--new issues section-->

---
<details><summary><b>New Issues (6)</b></summary>
Checkmarx found the following issues in this Pull Request

#|Severity|Issue|Source File / Package|Checkmarx Insight
---|---|---|---|---
1<!--1 RDszRftw7VNYr9BVQ%2FI6rq9wafI%3D IaC-Security-->|![LOW](https://cdn.ast.checkmarx.net/integrations/severity/Low.png)|Unpinned Actions Full Length Commit SHA|/release\.yml: [28](https://github.com/xinaesthete/viv/blob/deck_9.2/.github/workflows/release.yml#L28)|<details><summary>details</summary>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release\. Pinning to a particular SHA help...</details>
2<!--2 ROHUrGuC7%2FbpgztEfKE61sVRKcs%3D IaC-Security-->|![LOW](https://cdn.ast.checkmarx.net/integrations/severity/Low.png)|Unpinned Actions Full Length Commit SHA|/release\.yml: [34](https://github.com/xinaesthete/viv/blob/deck_9.2/.github/workflows/release.yml#L34)|<details><summary>details</summary>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release\. Pinning to a particular SHA help...</details>
3<!--3 i8C2rsuFnQnDHDUYjgS13iHQ2gE%3D IaC-Security-->|![LOW](https://cdn.ast.checkmarx.net/integrations/severity/Low.png)|Unpinned Actions Full Length Commit SHA|/ci\.yml: [39](https://github.com/xinaesthete/viv/blob/deck_9.2/.github/workflows/ci.yml#L39)|<details><summary>details</summary>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release\. Pinning to a particular SHA help...</details>
4<!--4 MxsKRkXxd5D8%2BRVx8m7UDjDP5gk%3D IaC-Security-->|![LOW](https://cdn.ast.checkmarx.net/integrations/severity/Low.png)|Unpinned Actions Full Length Commit SHA|/ci\.yml: [21](https://github.com/xinaesthete/viv/blob/deck_9.2/.github/workflows/ci.yml#L21)|<details><summary>details</summary>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release\. Pinning to a particular SHA help...</details>
5<!--5 gb%2FHwjHzuwKj77DuUKycbQlEbEE%3D IaC-Security-->|![LOW](https://cdn.ast.checkmarx.net/integrations/severity/Low.png)|Unpinned Actions Full Length Commit SHA|/ci\.yml: [32](https://github.com/xinaesthete/viv/blob/deck_9.2/.github/workflows/ci.yml#L32)|<details><summary>details</summary>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release\. Pinning to a particular SHA help...</details>
6<!--6 VrVF2AE5ULFEQVF%2FcIEO3lpoAjY%3D IaC-Security-->|![LOW](https://cdn.ast.checkmarx.net/integrations/severity/Low.png)|Unpinned Actions Full Length Commit SHA|/pages\.yml: [26](https://github.com/xinaesthete/viv/blob/deck_9.2/.github/workflows/pages.yml#L26)|<details><summary>details</summary>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release\. Pinning to a particular SHA help...</details>

</details>

<!--new issues section--><!--footer section-->

---
Use @Checkmarx to interact with Checkmarx PR Assistant. 
*Examples:*
`@Checkmarx how are you able to help me?`
`@Checkmarx rescan this PR`
<!--footer section-->

ilan-gold commented on packages/extensions/src/additive-colormap-3d-extensions/additive-blend-extension.js:
```diff
@@ -3,10 +3,10 @@ import BaseExtension from './base-extension';
 
 const _RENDER = `\
-  float intensityArray[6] = float[6](intensityValue0, intensityValue1, intensityValue2, intensityValue3, intensityValue4, intensityValue5);
+  float intensityArray[NUM_CHANNELS] = float[NUM_CHANNELS](intensityValue0, intensityValue1, intensityValue2, intensityValue3, intensityValue4, intensityValue5);
```
Yes, this is great!

xinaesthete replied:
Or rather... the syntax I mentioned works ok, but the new

```
uniform xrLayerUniforms {
  vec2 contrastLimits<VIV_CHANNEL_INDEX>;
} xrLayer;
```

I'm struggling with.

ilan-gold reviewed (commented):
I think it's reasonable to ask in `deck.gl` slack or on the github what the implications of a custom `ShaderAssembler` would be.  I really couldn't know off hand.  As you can probably tell, I did a lot of this by just looking at the underlying code and ballparking what would be needed.  The upside is it works.  The downside is that it has a bunch of private accesses. At the time, I was under paper pressure :/

I don't feel too strongly about how off-the-beaten-path this goes as long as it works.    We advertise breaking changes on the minor version so I'm also not worried about any breaking changes.  I'll try to download the deck.gl slack on Monday again.

xinaesthete commented:
@ilan-gold Ib already expressed some enthusiasm about viv updating deck versions and I'm sure he'll be happy to give some more feedback at some point. I'll add a pointer to this PR over there now.

xinaesthete commented on packages/extensions/src/additive-colormap-extension/additive-colormap-extension.js:
```diff
@@ -71,6 +77,8 @@ const AdditiveColormapExtension = class extends LayerExtension {
 
   updateState({ props, oldProps, changeFlags, ...rest }) {
+    // does any of this existing logic need to change?
+    // (like, is there ever more than one model?)
```
Yes I don't anticipate a problem there... In some places I'm looping over `this.getModels()` and it may be that it would be somewhat safer to do that a bit more consistently, but I'm pretty sure it isn't likely to matter in foreseeable future. Slight balance between consistency and avoiding changes where there isn't a solid reason for it.

xinaesthete committed:
mostly comments and inactive code for attempting contrastLimits UBO

xinaesthete committed:
use channel index template for volume shader

xinaesthete committed:
use imported template string in xr-fragment

xinaesthete committed:
start trying to refactor towards tagged-template for shaders

xinaesthete committed:
add missing MAX_CHANNELS import in bitmap-layer.js

xinaesthete committed:
biome fix, some comments

xinaesthete committed:
biome fix

xinaesthete commented:
n.b. I've made all the CI pass and will try to keep commits clean WRT that, but bear in mind that I don't think the configuration is properly working for `'@deck.gl/test-utils'` as noted https://github.com/hms-dbmi/viv/issues/921 so we need to be extra careful to manually verify that layers/extensions etc do actually work correctly.

xinaesthete committed:
move VIV_CHANNEL_INDEX_PLACEHOLDER to constants

xinaesthete committed:
more consistent application of VIV_CHANNEL_INDEX_PLACEHOLDER

xinaesthete committed:
Make avivator use value of MAX_CHANNELS from constants package. Change to 10 for testing.

xinaesthete committed:
biome fix

xinaesthete committed:
make tests pass with 10 channels

ilan-gold commented:
> Which is fine - but in other cases where we (or some hypothetical user with a custom extension) aren't able to loop over an array

So a different option would be to do our best, be super clear about what Viv does with its pipeline, and then just allow people to provide their own shaders completely (or almost completely, maybe just `main` plus custom uniforms).  Does that seem reasonable?

xinaesthete committed:
use string rather than Symbol for VIV_CHANNEL_INDEX_PLACEHOLDER

xinaesthete committed:
remove unused broken vivShader function

xinaesthete committed:
clean up and document shader line-expansion

xinaesthete committed:
fix typo "newContrastLimits.lengths"

xinaesthete committed:
wip expandShaderModule helper for uniform declarations

xinaesthete committed:
wip documentation

xinaesthete committed:
Refactor shader module handling to support per-channel uniform declarations, apply to contrast limit processing in xr-layer. Updated expandShaderModule to expand both uniform types and shader code. Adjusted XRLayer to utilize new channel intensity module structure.

xinaesthete committed:
complete refactor to UBO.
move VivShaderAssembler to extensions package, add NUM_PLANES template.
faff with extensions build settings in package.json.

xinaesthete committed:
remove redundant draw implementation in xr-layer

ilan-gold commented:
Awesome @xinaesthete - my biggest concern is definitely just that everythign is 100% clear.  If it's a small change or something subtle that most people won't notice, that's ok.  Re-request review once you're ready and I'll have a look!

ilan-gold commented:
> Probably once this is in a better state it might make sense to put the actual varying of number of channels at runtime into a separate PR, but I think I'm reasonably satisfied that the API for that is fairly stable and working.

That seems reasonable!

xinaesthete commented:
I might also be able to make a more detailed example of a moderately complex shader extension in the context of a spatialdata.js blog post with interactive sample.

xinaesthete commented:
> Regarding the shader assembler, I think it will also need some unit tests that check the final assembled shader string.

Yes - this is where having a test runner with an actual GL context will be required. Related - I had bugs with rarely-used (non-viv) extensions in MDV after a deck update. The shader extension code fragments referred to the old uniform names rather than the newly namespaced ones, so any test that was trying to actually compile the full shaders would've been able to surface it, but the feature wasn't much used so took a while for me to notice (and then it presents with a very glitchy debug UI at runtime - extra glitchy in MDV but not something you'd want to expose to users at the best of times).

So - this is something that needs to be very clear in the documentation (I think it's mentioned, but that documentation isn't in its final form), if you have any custom extensions, as well as changing your own declarations you need to be careful about which things in the surrounding code may have changed. I was obviously conscious of the related issues as I did that update in MDV but managed to miss that particular case.

I think we should make another PR to sort out that aspect of the test configuration, and then we can rebase/merge that into here. I don't mind looking into it if you like.

xinaesthete committed:
fix intermittent overflow jitter #863 for avivator deckgl-wrapper

xinaesthete committed:
add .pnpm-store to .gitignore

xinaesthete committed:
update deck.gl and luma.gl to 9.2.6, adding widgets and webgpu packages to satisfy unmet peer warnings.

xinaesthete committed:
some minor cleanup, better doc comment and comments on work-in-progress

xinaesthete committed:
improve colormap handling in XRLayer and AdditiveColormapExtension. Updated XRLayer to detect colormap changes and trigger model recreation accordingly. Removed model management from AdditiveColormapExtension to streamline responsibilities. Added comments for future review on colormap-related issues.

xinaesthete committed:
add null colormap to defaultProps in xrlayer to hypothetically help change detection

xinaesthete commented:
Currently on the roadmap possibly for luma.gl 9.4

https://github.com/visgl/luma.gl/issues/2408 

> Update shader module type system to handle arrays and structs

xinaesthete committed:
VivLayerExtension for shader management across extensions.
VivShaderAssembler still used mostly to avoid compile errors with NUM_CHANNELS in hook declaration.
Dynamic channel and plane count handling. Refactor shader module expansion to utilize new VivLayer interface.

xinaesthete committed:
make sure that bindings match changed channel count, add normalizeTextureBindings helper

xinaesthete committed:
Merge remote-tracking branch 'upstream/main' into deck_9.2

xinaesthete committed:
update deck.gl to 9.2.9

xinaesthete committed:
fix lint errors

xinaesthete committed:
address failing tests.
props.selections may be undefined in xr-layer.
private deckGLView._height etc no longer used.
ImageLayer implements getNumChannels() and getNumPlanes()

xinaesthete committed:
fix tests, guard against undefined props.selections in bitmap-layer

xinaesthete committed:
expandShaderModule adds NUM_CHANNELS and NUM_PLANES defines, avoiding edge-cases where code-paths missed them.

xinaesthete committed:
remove redundant NUM_PLANES/NUM_CHANNELS defines

xinaesthete committed:
BitmapLayer state management fixed to work with test_rgb.ome.tiff, more defensive/no mipmap

xinaesthete committed:
add glsl-parser to root devDepencies for test purposes

xinaesthete committed:
(llm) tests for shader-assembly

xinaesthete committed:
handle 0 channels gracefully, add tests

xinaesthete committed:
format test files

ilan-gold commented on packages/extensions/src/additive-colormap-extension/additive-colormap-extension.js:
```diff
@@ -72,26 +79,22 @@ const AdditiveColormapExtension = class extends LayerExtension {
-      }
+    // When colormap changes, getShaders() returns different modules, which triggers
+    // extensionsChanged in the layer, causing the layer to recreate models.
+    // We should NOT manage models here - that's the layer's responsibility.
```
Could you expand on this a bit? this seems like idiomatic `deck` to put model handling inside `updateState` for a layer extension https://deck.gl/docs/developer-guide/custom-layers/layer-extensions

ilan-gold commented on packages/layers/src/xr-layer/xr-layer.js:
Ok so we definitely need this.  I think we need to open an issue with luma.gl if this is going away.

ilan-gold commented on packages/layers/src/bitmap-layer.js:
```diff
@@ -135,22 +138,68 @@ const BitmapLayer = class extends CompositeLayer {
     super.initializeState(args);
   }
 
+  updateState({ props, oldProps }) {
```
Why is this needed now? Because of some `alpha` thing that broke?

ilan-gold reviewed (commented):
Nice, overall this is much simpler than I thought it would be.  You did an awesome job.

The two big things:

1. For me, I need to go line-by-line in some of the utility functions closer - this was more a high-level review
2. The docs should be more compact in general. I tried to get started on that but I think you know that already given the comments you put there.  It'll help me review though to have something more polished :)

xinaesthete commented on sites/docs/src/CUSTOM_SHADERS.md:
```diff
@@ -1,19 +1,364 @@
 
+> **WARNING - Breaking Changes**: If you have custom extensions, see the [Migration Guide](#migration-guide) below for important changes to support dynamic channel counts and deck.gl v9 UBO.
+
 Viv's shaders in 2D can be modified via [deck.gl shader hooks](https://deck.gl/docs/developer-guide/custom-layers/writing-shaders#standard-shader-hooks). The `LensExtension`, `AdditiveColormapExtension`, and `ColorPaletteExtesnion` (default) in Viv [implement shader hooks](https://github.com/hms-dbmi/viv/tree/main/src/extensions). Implementing your own shader hook requires extending the [standard ...
```
To be clear – UBO as such is inherently required for any uniforms in the new version of deck; so the distinction is whether our extra pre-processing code-generation is to be used. If I'd been able to see a good way to separate those things I'd have made a smaller PR that really was just focussed on UBO, but since it coincided with difficulties in using arrays in the way that viv had previously done, this was 

So, of course any changes necessary to make extensions compatible with the new deck.gl are unavoidable.

Old extensions, even not using channel/plane preprocessor related stuff, are likely to be broken if they
- declare any uniforms
- happen to reference some other uniforms etc, like standard deck stuff that has had changed declarations (these could already have been triggered by previous updates)
- rely on any viv-hooks that have changed - but this should only be true of things making use of channels etc.

You should be able to make a `ShaderExtension` and have it implement `getShaders()` without thinking about the strange application of `this` etc, and I believe that as long as it doesn't do anything that would be incompatible with deck itself (which will itself also use `extension.getShaders.call(layer, extension)` internally) it should work fine.

So perhaps it would be worth verifying this hypothesis, but I believe that if we rephrase
> You must inherit from `VivLayerExtension` if you want to do UBO stuff, otherwise you can still use the old machinery.

more along the lines of

> Inheriting from `VivLayerExtension` is strongly advised particularly in cases where code-generation is to be used, otherwise it should be possible to use a `LayerExtension` as long as it is migrated along the lines required by changes in deck.gl.

There do exist extensions within the code that this could apply to, so we could try changing those back to verify.

xinaesthete committed:
fix ColorPaletteExtension mutate_color implementation, numChannels from selections length

xinaesthete committed:
minor change to lodash version because of ci error

xinaesthete committed:
move _setNewTexturesFromLoadThisFrame into helper

xinaesthete committed:
pin pnpm version in CI to match package.json

xinaesthete committed:
documentation work-in-progress

xinaesthete committed:
Merge branch 'main' into deck_9.2

xinaesthete committed:
set lodash version constraint to match upstream again

xinaesthete committed:
update pnpm-lock

xinaesthete committed:
tests for extensions

xinaesthete committed:
lint

xinaesthete commented on packages/extensions/src/color-palette-extension/color-palette-module.js:
```diff
@@ -1,36 +1,47 @@
+  );
+  for(int i = 0; i < NUM_CHANNELS; i++) {
+    rgb += max(0.0, min(1.0, intensity)) * vec3(colors[i]);
+  }
```
Tests now added for this.

xinaesthete committed:
remove redundant uniformTypes in getRenderingAttrs
add helper function for better type in test

xinaesthete committed:
extension tests use getShaders rather than expandShaderModule

xinaesthete committed:
lint

xinaesthete committed:
update documentation

xinaesthete committed:
remove redundant test of getRenderingAttrs uniformTypes which no longer exist

xinaesthete committed:
fix memory errors #949 by disabling pool again

xinaesthete committed:
avoid mutating props, make sure to call super.updateState/finalizeState in bitmap-layer

xinaesthete committed:
remove numTextures which was always equal to numChannels

xinaesthete committed:
fix example VivLayerExtension in doc

xinaesthete committed:
change extension example to spread rest props in updateState

xinaesthete commented:
There seems to be a fairly major performance degradation with volume rendering currently – I would expect it to be faster than before as long as it's rendering a lower number of channels, and I believe this was the case previously, so I'll try to figure out what's going on there.

xinaesthete commented:
> There seems to be a fairly major performance degradation with volume rendering currently – I would expect it to be faster than before as long as it's rendering a lower number of channels, and I believe this was the case previously, so I'll try to figure out what's going on there.

**edit:** I have noticed that it seems much worse for camera interactions than tweaking contrast-limits. At least with my local changes there are intermittent pauses when moving the camera, so this is something to investigate; seems like the shader details are not so relevant. I might post on the deck.gl Slack.

Also, I think I've been seeing some stutters in 2D as well, but not on the same level.

### older comments
I've realised I've got channel-expansion to arrays inside the hot-path of the raymarch loop in the volume shader... whoops... ~~that'd likely explain some performance degradation~~. I think this is actually all pretty trivial these days and I'm not seeing much change from hoisting things out of loops etc; I expect somewhere up the GLSL compiler stack it's able to do that automatically.

~~It seems that there are calls being made to `bufferSubData` every frame which surely shouldn't be necessary~~. Wasn't thinking straight there – this is the normal expected UBO updating, I don't think I've seen texture data being thrashed.

Hopefully it's just a case of tweaking `updateState` / `draw` etc.

xinaesthete committed:
call to getVivShaderTemplates binds this to make sure props are accessible, fixes bug with always selecting default colormap

xinaesthete committed:
using line-expansion rather than arrays and loops for 3d shaders

xinaesthete committed:
refactor xr-3d-layer particularly to avoid GC

ilan-gold commented:
@xinaesthete Apologies, was probably https://github.com/xinaesthete/viv/blob/deck_9.2/packages/viewers/src/VivViewer.jsx#L73-L75 which doesn't actually seem to work on everyone's machines/site builds.

I reverted it in #948 but am waiting on @keller-mark there.  Maybe I should just revert globally.  Sorry for the trouble!

xinaesthete committed:
Merge branch 'main' into deck_9.2

xinaesthete commented on sites/avivator/src/index.css:
```diff
@@ -12,3 +12,8 @@ body {
+
+/* this aleviates an intermittent issue with glitchy overflow oscillation */
+#deckgl-wrapper {
+  overflow: hidden;
```
n.b. this is orthogonal to the actual deck.gl changes and other refactor, but at some point that was suddenly manifesting in a way that was pretty intolerable to work with so I added it here. Not sure if this is in any way related to the https://github.com/hms-dbmi/viv/pull/948 you mentioned @ilan-gold?

xinaesthete replied:
arrrgh... I just looked at it in local dev and the dreaded oscillating overflow is back.

edit: I think I was actually confused about which server I was looking at, and this was from `main`.

ilan-gold replied:
It could be - this one I haven't seen but if you split it off into its own PR and/or try it out with the scale bar thing, lmk if that works!

xinaesthete replied:
> It could be - this one I haven't seen but if you split it off into its own PR and/or try it out with the scale bar thing, lmk if that works!

It's horrendous when it happens.... 

https://github.com/user-attachments/assets/95602ac6-84b0-4a9e-9c48-902db5f141e1

xinaesthete replied:
https://github.com/hms-dbmi/viv/pull/953

xinaesthete committed:
lint & typo fix

xinaesthete committed:
Squashed commit of the following:

commit 91560abba06e4d14aec5374bab0fbc9cf3bea017
Author: Ilan Gold <ilanbassgold@gmail.com>
Date:   Thu Mar 19 17:22:12 2026 +0100

    fix: remove debounce (#952)

xinaesthete commented on .github/workflows/ci.yml:
```diff
@@ -19,6 +19,8 @@ jobs:
       - uses: actions/checkout@v4
       - uses: pnpm/action-setup@v4
+        with:
+          version: 9.5.0
```
I changed these because at one point I didn't realise that CI on the branch was running on an automatic merge with `main` and I was trying to figure out why CI was failing when everything worked locally. `package.json` has `"packageManager": "pnpm@9.5.0"`, there may be something to be said for making sure that these are all tied to the same version, but the change isn't necessary to the PR (CI was breaking because of silently auto-merged `pnpm-lock.yml` I think), and if we make a separate change it'd probably also update to a newer version.

xinaesthete committed:
add changeset

keller-mark committed:
Minor cleanup

keller-mark committed:
Merge

keller-mark committed:
Format

xinaesthete commented:
Since there have been some more releases in the meantime, might be worth seeing if we can cleanly update to the latest deck/luma before merging. I don't think there's anything too critical there, though – hopefully the breaking changes are done now (at least until someone gets a WebGPU itch... which might happen in the not-so-distant future).

ilan-gold commented on packages/extensions/src/color-palette-extension/color-palette-extension.js:
```diff
@@ -18,33 +18,51 @@ const defaultProps = {
   }
 
-  draw() {
+  updateState({ props, oldProps, changeFlags, ...rest }) {
```
Nice this is much cleaner

ilan-gold commented on packages/extensions/src/color-palette-extension/color-palette-module.js:
```diff
@@ -1,36 +1,46 @@
 const DECKGL_MUTATE_COLOR = `\
 vec3 rgb = rgba.rgb;
-mutate_color(rgb, intensity0, intensity1, intensity2, intensity3, intensity4, intensity5);
+mutate_color(rgb, intensity, vTexCoord);
```
`vTexCoord` not needed here or? It seems they all have the same signature but I don't see any practical reason for that, no?

xinaesthete replied:
I think it's useful to give extensions flexibility to use this? I can try to think a bit more of something more concrete if you like.

xinaesthete replied:
Having said that, I think in the past I ran into issues with actually accessing sample data from an extension because of declaration order.

ilan-gold replied:
My point, I think, is that this `mutate_color` is internal - users creating extensions would have their own `mutate_color` variant that could operate independent of the `vTexCoord`

xinaesthete replied:
Yes I just opened up the code to look at this in context again and patched it without any discernible issue.

ilan-gold commented on packages/extensions/tests/fixtures/extensions-shader-modules.fixtures.ts:
```diff
@@ -0,0 +1,123 @@
+  return vec4(color, (color == transparentColor && useTransparentColor) ? 0. : opacity);
+}
+
+void mutate_color(inout vec3 rgb, float[NUM_CHANNELS] intensity, vec2 vTexCoord) {
```
No use of `vTexCoord`?

ilan-gold reviewed (approved):
Nice! @keller-mark I would just cut a release for this and then another for the next upgrade, no?

keller-mark commented:
Yes I think it would be best to maximize the amount of compatible deckGL minor versions, avoiding skipping over them if we can avoid it

xinaesthete committed:
remove unused vTexCoord parameter in color-palette-module

xinaesthete commented:
> Yes I think it would be best to maximize the amount of compatible deckGL minor versions, avoiding skipping over them if we can avoid it

I just locally changed to deck ~9.3.2 & luma ~9.3.3 and got this warning:

```
 WARN  Issues with peer dependencies found
.
└─┬ @deck.gl/test-utils 9.3.2
  └── ✕ unmet peer vitest@^4.0.18: found 3.2.4
```

This implies that it'll need a Vite upgrade as well, which would be good to do but has potential to be problematic. Probably better to do that in a separate PR.

Having said that, it does appear to run the tests ok despite the warning. I might just try bumping vite & vitest to see whether all the general build config etc survives... but really it probably is better for another PR just in case something slips through unnoticed.

xinaesthete commented:
@keller-mark I think we could merge this as-is and then I update the other PR updating Vite et al to also do the latest version of deck stack.

ilan-gold commented:
Thanks!

xinaesthete commented:
@ilan-gold I've just seen there is actually something of a regression in how the scalebar is transformed, should make another issue...

**edit:** not seeing this now going back to this branch is it was so I think it's just something that needs fixing in the other PR.

ilan-gold commented:
@xinaesthete Yeah, I did some testing for 15 min on my end, didn't see anything.

ilan-gold commented:
And the demo is back to be very smooth for me!!!!

xinaesthete commented:
@ilan-gold fixed in the other branch now
